components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,328 @@
+    /// `start`: inclusive, `end`: exclusive
+    ///
+    /// Returns an error if the range is invalid (out of order and out of bounds).
+    ///
+    /// Example Call: `UnicodeSet::from_range(&0, &15)`
+    pub fn from_range(start: &u32, end: &u32) -> Result<UnicodeSet, Box<dyn Error>> {
+        if start > end {
+            return Err("Range is out of order".into());
+        }
+        if start < &CODEPOINT_MIN || end > &CODEPOINT_MAX {
+            return Err("Range is out of bounds".into());
+        }
+        Ok(UnicodeSet {
+            inv_list: vec![*start, *end],
+        })
+    }
+
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, CODEPOINT_MAX + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, BMP_MAX + 1],
+        }
+    }
+
+    /// Returns an `Iter` of start and stop `u32` points of the UnicodeSet
+    pub fn iter(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Returns the cardinality of the UnicodeSet
+    ///
+    ///
+    pub fn size(&self) -> Result<usize, Box<dyn Error>> {
+        if self.inv_list.len() < 2 {
+            return Err("UnicodeSet length < 2".into());
+        }
+        let end_point_sum: u32 = self.iter().skip(1).step_by(2).sum::<u32>();
+        let start_point_sum: u32 = self.iter().step_by(2).sum::<u32>();
+        Ok((end_point_sum - start_point_sum) as usize)
+    }
+
+    /// Returns whether or not the UnicodeSet is empty
+    pub fn is_empty(&self) -> bool {
+        self.inv_list.len() < 2 // unsure if this is appropriate definition of just self.inv_list.is_empty()
+    }
+
+    /// Wrapper for contains conditions closures
+    fn contains<C>(&self, query: &u32, condition: C) -> bool
+    where
+        C: Fn(usize) -> bool,
+    {
+        match self.inv_list.binary_search(query) {
+            Ok(pos) => {
+                if pos % 2 == 0 {
+                    condition(pos)
+                } else {
+                    false
+                }
+            }
+            Err(pos) => {
+                if pos % 2 != 0 && pos < self.inv_list.len() {
+                    condition(pos)
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    /// Checks to see the query is in the UnicodeSet
+    ///
+    /// Runs a binary search in `O(log(n))` where `n` is the number of start and end points
+    /// in the set using `std::vec::Vec` implementation
+    ///
+    /// Example: `contains_point(&10)`
+    pub fn contains_point(&self, query: &u32) -> bool {
+        let condition_closure = |_: usize| -> bool { true };
+        self.contains(query, condition_closure)
```

Comment:
I agree, I was unsure if the closure would work without specifying the `usize` since the parameter requires it but you're right, the compiler won't complain.